[syntax-errors] Document behavior of `global` declarations in `try` nodes before 3.13

ntBre opened the pull request:
Summary
--

This PR extends the documentation of the `LoadBeforeGlobalDeclaration` check to specify the behavior on versions of Python before 3.13. Namely, on Python 3.12, the `else` clause of a `try` statement is visited before the `except` handlers:

```pycon
Python 3.12.9 (main, Feb 12 2025, 14:50:50) [Clang 19.1.6 ] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a = 10
>>> def g():
...     try:
...         1 / 0
...     except:
...         a = 1
...     else:
...         global a
...
>>> def f():
...     try:
...         pass
...     except:
...         global a
...     else:
...         print(a)
...
  File "<stdin>", line 5
SyntaxError: name 'a' is used prior to global declaration

```

The order is swapped on 3.13 (see [CPython#111123](https://github.com/python/cpython/issues/111123)):

```pycon
Python 3.13.2 (main, Feb  5 2025, 08:05:21) [GCC 14.2.1 20250128] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a = 10
... def g():
...     try:
...         1 / 0
...     except:
...         a = 1
...     else:
...         global a
...
  File "<python-input-0>", line 8
    global a
    ^^^^^^^^
SyntaxError: name 'a' is assigned to before global declaration
>>> def f():
...     try:
...         pass
...     except:
...         global a
...     else:
...         print(a)
...
>>>
```

The current implementation of PLE0118 is correct for 3.13 but not 3.12: [playground](https://play.ruff.rs/d7467ea6-f546-4a76-828f-8e6b800694c9) (it flags the first case regardless of Python version). 

We decided to maintain this incorrect diagnostic for Python versions before 3.13 because the pre-3.13 behavior is very unintuitive and confirmed to be a bug, although the bug fix was not backported to earlier versions. This can lead to false positives and false negatives for pre-3.13 code, but we also expect that to be very rare, as demonstrated by the ecosystem check (before the version-dependent check was reverted here).

Test Plan
--

N/a

ntBre commented:
I just realized that this might need to be gated behind preview since it modifies an existing, stable lint rule. Alternatively, I could just use a different `SemanticSyntaxErrorKind` here.

ntBre commented:
Oh, there's a false positive anyway, moving back to draft.

dhruvmanila commented:
> I think `u32` should be huge overkill for the scope depth tracking, so I'm more than happy to drop the size of that to a more reasonable level, but I wasn't sure how low to go. Is `u8` enough?

`u32` should be fine, we use the same for `ScopeId`:

https://github.com/astral-sh/ruff/blob/058439d5d3f65db515d69359c99838d60776f45e/crates/ruff_python_semantic/src/scope.rs#L188-L194

dhruvmanila commented:
> On Python 3.12, the `else` clause is visited before the `except` handlers:

This is confusing, do we have a source like a changelog entry where the related change is mentioned for context? It might also be useful to add that as comment in the source code.

dhruvmanila commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -125,6 +128,37 @@ impl SemanticSyntaxChecker {
+                is_star: _,
+            }) => {
+                let mut visitor = TryExceptVisitor::new(ctx);
+                for stmt in body {
+                    SourceOrderVisitor::visit_stmt(&mut visitor, stmt);
+                }
```
We should use the `visit_body` method instead in here, orelse, handlers, and finalbody.

dhruvmanila replied:
except for the "handlers" loop

MichaReiser replied:
I think the easiest here is to call `walk_stmt(&visitor, stmt)` because this is an exact reimplementation of visiting a try statement.

MichaReiser replied:
You're right, we traverse into function statements but we increase the `scope` counter. I have to take another look to understand what the scope counter is even used for but I suspect that different rules apply for names and global statements inside nested functions, classes and it's unclear to me if that also applies to lambdas (can't have any global statements but it may have identifiers)

ntBre commented:
> > On Python 3.12, the `else` clause is visited before the `except` handlers:
> 
> This is confusing, do we have a source like a changelog entry where the related change is mentioned for context? It might also be useful to add that as comment in the source code.

This is the [bug report](https://github.com/python/cpython/issues/111123) Alex linked to in #11934. I'll look for anything else and at least link to this in the enum variant docs!

dhruvmanila commented:
> This is the [bug report](https://github.com/python/cpython/issues/111123) Alex linked to in #11934. I'll look for anything else and at least link to this in the enum variant docs!

I think just linking to the CPython issue should be enough.

dhruvmanila commented:
(I need to head out now, I'll continue with the review later in the evening as I need to look at it a bit more closely to understand.)

MichaReiser commented:
> I just realized that this might need to be gated behind preview since it modifies an existing, stable lint rule. Alternatively, I could just use a different SemanticSyntaxErrorKind here, which seems a little easier. On the other hand, this could probably be considered a bug in the current rule implementation.

Can you tell me more in how the new implementation diverges from the existing one?

MichaReiser commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -125,6 +128,37 @@ impl SemanticSyntaxChecker {
+                finalbody,
+                is_star: _,
+            }) => {
+                let mut visitor = TryExceptVisitor::new(ctx);
```
Using a custom visitor here is probably fine because try statements aren't too deeply nested (they rarely contain nested functions) but this is now a case where the visitor ends up visiting entire suites multiple times: 

1. In the `TryExcept` visitor
2. The outer visitor

I've a weak preference to avoid the double visitation unless it increases complexity by a lot

ntBre replied:
I'll give this a try (no pun intended). The nice thing about the visitor is that it let me control the order in which things are seen, and I could just swap the order of the visits on different Python versions. I think that order-dependence might be hard to capture otherwise, but I'll play with it a bit.

The easiest thing to do would be to move that reordering into the parent visitor, but that seems too intrusive.

MichaReiser commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -1280,6 +1314,143 @@ impl<'a, Ctx: SemanticSyntaxContext> MatchPatternVisitor<'a, Ctx> {
+
+    fn visit_stmt(&mut self, stmt: &Stmt) {
+        match stmt {
+            Stmt::FunctionDef(_) | Stmt::ClassDef(_) => {
```
What about nested `try` statements?

ntBre replied:
I don't think `try` introduces a new scope:

```pycon
>>> try:
...     x = 1
... except:
...     pass
...
>>> x
1
>>>
```

I think only functions and classes do. Is that what you were asking?

MichaReiser commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -1280,6 +1314,143 @@ impl<'a, Ctx: SemanticSyntaxContext> MatchPatternVisitor<'a, Ctx> {
+    }
+
+    fn visit_expr(&mut self, expr: &Expr) {
+        if let Expr::Name(ast::ExprName { range, id, ctx: _ }) = expr {
```
What about lambda expressions?

MichaReiser replied:
Yes, the visitor traverses into lambdas. The question is whether it should (e.g. we don't traverse into nested functions or classes)

MichaReiser commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -1280,6 +1314,143 @@ impl<'a, Ctx: SemanticSyntaxContext> MatchPatternVisitor<'a, Ctx> {
+
+    fn visit_expr(&mut self, expr: &Expr) {
+        if let Expr::Name(ast::ExprName { range, id, ctx: _ }) = expr {
+            self.add_identifier(id.clone(), *range);
```
I find it unfortunate that we're collecting all identifiers here, even if there isn't a single global statement in the program (which should be the vast majority of programs). 

It also has the downside that globals handling is now handled here but also something that we expect the caller to handle (`context.global(name)`). Could we instead make use that `context` already provides a `global` method or extend the `context` interface to avoid doing this work twice?

MichaReiser commented on crates/ruff_python_parser/src/semantic_errors.rs:
```diff
@@ -1280,6 +1314,143 @@ impl<'a, Ctx: SemanticSyntaxContext> MatchPatternVisitor<'a, Ctx> {
+struct TryExceptVisitor<'a, Ctx> {
+    /// Context used for emitting errors.
+    ctx: &'a Ctx,
+    identifiers: FxHashMap<ast::name::Name, Ident>,
```
We should store `name` references here (you have the `'a` lifetime anyway)

MichaReiser reviewed (commented):
We should add a few more tests and explore if there's a way to avoid duplicating the globals handling between the semantic syntax checker and the caller.

ntBre commented:
> Can you tell me more in how the new implementation diverges from the existing one?

The divergence is just the version-dependence of the branch visiting order. From the summary:

> The current implementation of PLE0118 is correct for 3.13 but not 3.12: [playground](https://play.ruff.rs/d7467ea6-f546-4a76-828f-8e6b800694c9) (it flags the first case regardless of Python version).

Phrasing it that way, it clearly sounds like a bug, but I think the old behavior was so unintuitive that probably nobody runs into it

ntBre commented:
Based on our discussion on Discord, I reverted all of the code changes here and simply extended the rule variant documentation to reflect our decision to enforce the 3.13 clause ordering on all Python versions.

I initially planned to keep the inline tests to demonstrate this, but they would now rely on a proper implementation of `SemanticSyntaxContext::global` for the test context to work, so I just deleted them too. We could add a linter test for PLE0118 showing this, though, if we wanted.

ntBre commented:
The `cargo-shear` failure was an issue in downloading/building it, not an extra dependency. I'll try to rerun it

MichaReiser reviewed (approved):
Thanks and sorry for not raising this sooner

ntBre commented:
No problem at all, I thought it was a useful discussion!
